flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java

```diff
@@ -174,27 +215,238 @@ public static void printAsTableauForm(
         printWriter.flush();
     }
 
-    public static String[] rowToString(Row row) {
-        return rowToString(row, NULL_COLUMN, false);
+    public static String[] rowToString(
+            Row row, ResolvedSchema resolvedSchema, ZoneId sessionTimeZone) {
+        return rowToString(row, NULL_COLUMN, false, resolvedSchema, sessionTimeZone);
     }
 
-    public static String[] rowToString(Row row, String nullColumn, boolean printRowKind) {
+    public static String[] rowToString(
+            Row row,
+            String nullColumn,
+            boolean printRowKind,
+            ResolvedSchema resolvedSchema,
+            ZoneId sessionTimeZone) {
         final int len = printRowKind ? row.getArity() + 1 : row.getArity();
         final List<String> fields = new ArrayList<>(len);
         if (printRowKind) {
             fields.add(row.getKind().shortString());
         }
         for (int i = 0; i < row.getArity(); i++) {
             final Object field = row.getField(i);
+            final LogicalType fieldType =
+                    resolvedSchema.getColumnDataTypes().get(i).getLogicalType();
             if (field == null) {
                 fields.add(nullColumn);
             } else {
-                fields.add(StringUtils.arrayAwareToString(field));
+                fields.add(
+                        StringUtils.arrayAwareToString(
+                                normalizeTimestamp(field, fieldType, sessionTimeZone)));
             }
         }
         return fields.toArray(new String[0]);
     }
 
+    /**
+     * Normalizes field that contains TIMESTAMP and TIMESTAMP_LTZ type data.
+     *
+     * <p>This method supports array type and nested type.
+     */
+    private static Object normalizeTimestamp(
```

Comment:
The logic is quite huge, would be better to extract the logic into a `DataStructureStringWriter`, similar to `DataStructureConverter`. We also have another issue FLINK-21456 discussed about normalizing all data types, not only timestamps.  I think this can be good start point.

Reply:
After discussion with Jark,  we think we can refactor this in the future by adding a string representation for all internal data structure (i.e. RowData), in this way we can reuse the string representation for output and planner codegen